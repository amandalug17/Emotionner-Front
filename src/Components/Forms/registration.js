import React, { useState, useRef } from "react";
import Form from "react-validation/build/form";
import Input from "react-validation/build/input";
import CheckButton from "react-validation/build/button";
import { isEmail } from "validator";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faUser} from '@fortawesome/free-solid-svg-icons'

import AuthService from "../../Services/auth.service";
/**
 * Requeried method verifies that all the fields are filled in
 * @param {*} value 
 */
const required = (value) => {
  if (!value) {
    return (
      <div className="alert alert-danger" role="alert">
        Este campo es requerido!
      </div>
    );
  }
};
/**
 * Email validation method, validates if an email is a valid format
 * @param {*} value 
 */
const validEmail = (value) => {
  if (!isEmail(value)) {
    return (
      <div className="alert alert-danger" role="alert">
        Este no es un email válido, porfavor ingrese una dirección valida.
      </div>
    );
  }
};
/**
 * Password validator, validates if the leght of the password is correct it 
 * must be between 6 and 40 characters
 * @param {*} value 
 */
const vpassword = (value) => {
  if (value.length < 6 || value.length > 40) {
    return (
      <div className="alert alert-danger" role="alert">
        La contraseña debe tener de 6 a 40 caracteres.
      </div>
    );
  }
};
/**
 * Registration Hook
 * @param {*} props 
 */

const Register = (props) => {
  const form = useRef();
  const checkBtn = useRef();
  /**
   * Variables
   */
  const [name, setName] = useState("");
  const [lastname, setLastname] = useState("");
  const [email, setEmail] = useState("");
  const [birthdate, setBirthdate] = useState("");
  const [ocupation, setOcupation] = useState("");
  const [premium, setPremium] = useState(false);
  const [password, setPassword] = useState("");
  const [successful, setSuccessful] = useState(false);
  const [message, setMessage] = useState("");

/**
 * Const values changes
 * @param {*} e 
 */
  const onChangeName = (e) => {
    const name = e.target.value;
    setName(name);
  };

  const onChangeLastName = (e) => {
    const lastname = e.target.value;
    setLastname(lastname);
  };

  const onChangeEmail = (e) => {
    const email = e.target.value;
    setEmail(email);
  };
  const onChangeBirthdate = (e) => {
    const birthdate = e.target.value;
    setBirthdate(birthdate);
  };
  const onChangeOcupation = (e) => {
    const ocupation = e.target.value;
    setOcupation(ocupation);
  };


  const onChangePassword = (e) => {
    const password = e.target.value;
    setPassword(password);
  };

  const handleRegister = (e) => {
    e.preventDefault();

    setMessage("");
    setSuccessful(false);

    form.current.validateAll();

    if (checkBtn.current.context._errors.length === 0) {
      AuthService.register(name,lastname,email,birthdate,ocupation,premium,password).then(
        (response) => {
          setMessage(response.data.message);
          setSuccessful(true);
        },
        (error) => {
          const resMessage =
            (error.response &&
              error.response.data &&
              error.response.data.message) ||
            error.message ||
            error.toString();

          setMessage(resMessage);
          setSuccessful(false);
        }
      );
    }
  };

  return (
    <div className="container">
      <div className="card card-signin my-5">
        <div className="card-body">
            <div className='form-icon'>
              <span ><FontAwesomeIcon icon={faUser} /></span>
            </div>
            <h5 className='card-title text-center text-uppercase'>Registrate</h5>
            <Form onSubmit={handleRegister} ref={form}>
              {!successful && (
                <div>
                  <div className="form-group">
                    <label htmlFor="name">Nombre</label>
                    <Input
                      type="text"
                      className="form-control"
                      name="name"
                      value={name}
                      onChange={onChangeName}
                      validations={[required]}
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="lastname">Apellido</label>
                    <Input
                      type="text"
                      className="form-control"
                      name="lastname"
                      value={lastname}
                      onChange={onChangeLastName}
                      validations={[required]}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="email">Email</label>
                    <Input
                      type="text"
                      className="form-control"
                      name="email"
                      value={email}
                      onChange={onChangeEmail}
                      validations={[required, validEmail]}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <Input
                      type="password"
                      className="form-control"
                      name="password"
                      value={password}
                      onChange={onChangePassword}
                      validations={[required, vpassword]}
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="birthdate">Fecha de Nacimiento</label>
                    <Input
                      type="date"
                      className="form-control"
                      name="birthdate"
                      value={birthdate}
                      onChange={onChangeBirthdate}
                      validations={[required]}
                    />
                  </div>

                  <div className="form-group">
                    <button className="btn btn-primary btn-block">Registrarse</button>
                  </div>
                </div>
              )}

              {message && (
                <div className="form-group">
                  <div
                    className={
                      successful ? "alert alert-success" : "alert alert-danger"
                    }
                    role="alert"
                  >
                    {message}
                  </div>
                </div>
              )}
              <CheckButton style={{ display: "none" }} ref={checkBtn} />
        </Form>
        </div>

        
      </div>
    </div>
  );
};

export default Register;