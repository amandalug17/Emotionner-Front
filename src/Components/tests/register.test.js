import registration from '../Forms/registration'
import React from 'react';
import { shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AuthService from "../../Services/auth.service";
import mockAxios from 'axios';

describe('test for registration form', () => {
    it ('renders correctly', () => {
        expect(<registration.Register />).toMatchSnapshot();
    })

    
    it('should call onChange with an imput value in name', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        component.find('.name1').dive().find('.form-control').simulate('change', {target: {
            value: 'Change function'
        }});
        expect(JSON.stringify(component)).toMatchSnapshot();
    })

    it('should call onChange with an imput value in lastname', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        component.find('.Lastname').dive().find('.form-control').simulate('change', {target: {
            value: 'Change function'
        }});
        expect(JSON.stringify(component)).toMatchSnapshot();
    })

    it('should call onChange with an imput value in email', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        component.find('.email').dive().find('.form-control').simulate('change', {target: {
            value: 'Change function'
        }});
        expect(JSON.stringify(component)).toMatchSnapshot();
    })

    it('should call onChange with an imput value in password', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        component.find('.password').dive().find('.form-control').simulate('change', {target: {
            value: 'Change function'
        }});
        expect(JSON.stringify(component)).toMatchSnapshot();
    })

    it('should call onChange with an imput value in Birthdate', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        component.find('.Bdate').dive().find('.form-control').simulate('change', {target: {
            value: 'Change function'
        }});
        expect(JSON.stringify(component)).toMatchSnapshot();
    })

    it('sends successfully data to Heroku', done => {

        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();

        mockAxios.post = jest.fn().mockResolvedValueOnce({});

        AuthService.register('name', 'lastname', 'email', 'birthdate', 'ocupation', 'premium', 'password').then(response => {
            expect(response).toEqual({
                data: {},
            });
        });
        
        expect(mockAxios.post).toHaveBeenCalledWith('https://emotionner.herokuapp.com/users/signup', {
            name : 'name',
            lastname: 'lastname',
            email : 'email',
            birthdate : 'birthdate',
            ocupation : 'ocupation',
            premium : 'premium',
            password : 'password'
        });
        
        expect(mockAxios.post).toHaveBeenCalledTimes(1);
        expect(consoleErrorSpy).not.toHaveBeenCalled();
        consoleErrorSpy.mockRestore();
        done();
    });

    it('should validate inputs to not be null', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        registration.required = jest.fn().mockResolvedValueOnce("");
        if(component.find('.name1').dive().find('.form-control').simulate('change', {target: {
            value: ''
        }})){
            registration.required("");
            expect(registration.required).toHaveBeenCalledTimes(1);
        } else {
            expect(registration.required).toHaveBeenCalledTimes(0);
        }

    })

    it('should validate email', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        registration.validEmail = jest.fn().mockResolvedValueOnce("");
        if(component.find('.email').dive().find('.form-control').simulate('change', {target: {
            value: 'emailNotValid'
        }})){
            registration.validEmail("");
            expect(registration.validEmail).toHaveBeenCalledTimes(1);
        } else {
            expect(registration.validEmail).toHaveBeenCalledTimes(0);
        }

    })

    it('should validate password', () => {
        configure({ adapter: new Adapter() })
        const component = shallow(<registration.Register/>);
        registration.vpassword = jest.fn().mockResolvedValueOnce("");
        if(component.find('.email').dive().find('.form-control').simulate('change', {target: {
            value: '1234'
        }})){
            registration.vpassword("");
            expect(registration.vpassword).toHaveBeenCalledTimes(1);
        } else {
            expect(registration.vpassword).toHaveBeenCalledTimes(0);
        }

    })

})