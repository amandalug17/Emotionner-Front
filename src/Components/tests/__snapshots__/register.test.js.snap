// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for registration form renders correctly 1`] = `<Register />`;

exports[`test for registration form should call onChange with an imput value in Birthdate 1`] = `"{}"`;

exports[`test for registration form should call onChange with an imput value in email 1`] = `"{}"`;

exports[`test for registration form should call onChange with an imput value in lastname 1`] = `"{}"`;

exports[`test for registration form should call onChange with an imput value in name 1`] = `"{}"`;

exports[`test for registration form should call onChange with an imput value in password 1`] = `"{}"`;
